### PR TITLE
rootless: add cli validator

### PR DESCRIPTION
**Pull request by giuseppe**
whenever the podman process is launched, it runs any file found in
    these directories:
    
    - /etc/containers/auth-scripts
    - /usr/libexec/podman/auth-scripts
    
    The current podman command line is passed as arguments to the
    process.
    
    If any of the processes fail, the error is immediately reported back
    from podman that exits with the same error code.
    
    [NO NEW TESTS NEEDED] requires a system-wide configuration.
    
    Signed-off-by: Giuseppe Scrivano <gscrivan@redhat.com>

#### Does this PR introduce a user-facing change?

```release-note
None
```

### CODE REVIEW

**Comment by rhatdan** [comment]
It would be nice to implement a test for this using a hidden environment variable for an auth directory.
LGTM

File: pkg/rootless/rootless_linux.c
Comment (Luap99) [inline pkg/rootless/rootless_linux.c:166]: In case of a error can we print a extra line such as `Error: external auth script failed`? If someone ever complains that podman is not working because some auth executable exits silently, it should be clear that this was not a podman error without running strace.

**Comment by giuseppe** [comment]
> This also needs documentation!

we left this undocumented because it was a customer request but we are not sure yet if we want to use/support it upstream

**Comment by rhatdan** [comment]
LGTM

**Comment by giuseppe** [comment]
the customer asked for these checks to happen on the client, for whoever is using the podman command, even in the case of invoking podman on a different host.

That is the reason why they are part of the rootless package

**Comment by Luap99** [comment]
> the customer asked for these checks to happen on the client, for whoever is using the podman command, even in the case of invoking podman on a different host.
> 
> That is the reason why they are part of the rootless package

OK if it is supposed to check the remote client as well fine but they don't happen on the client if you use the macos/windows version, so this is inconsistent behaviour.

And more importantly this really should not be forced on `pkg/bindings` users.

**Comment by giuseppe** [comment]
> > the customer asked for these checks to happen on the client, for whoever is using the podman command, even in the case of invoking podman on a different host.
> > That is the reason why they are part of the rootless package
> 
> OK if it is supposed to check the remote client as well fine but they don't happen on the client if you use the macos/windows version, so this is inconsistent behaviour.
> 
> And more importantly this really should not be forced on `pkg/bindings` users.

do these users really have something under `/etc/containers/auth-scripts` and  `/usr/libexec/podman/auth-scripts`?

**Comment by rhatdan** [comment]
Lets worry about handling this on Linux for now, and it should happen in client code.

**Comment by Luap99** [comment]
> > > the customer asked for these checks to happen on the client, for whoever is using the podman command, even in the case of invoking podman on a different host.
> > > That is the reason why they are part of the rootless package
> > 
> > 
> > OK if it is supposed to check the remote client as well fine but they don't happen on the client if you use the macos/windows version, so this is inconsistent behaviour.
> > And more importantly this really should not be forced on `pkg/bindings` users.
> 
> do these users really have something under `/etc/containers/auth-scripts` and `/usr/libexec/podman/auth-scripts`?

We cannot know that but just imagine someone writing a program which uses our bindings and then it starts executing the program in there and it can fail because of it. How is a user expected to know what is going on. There is no documentation, the only chance is to lookup the source code.

---

I am really confused what the purpose of this is. When they try to restrict the client what is preventing someone from just using curl or any other tool to open a socket and send http API query's to the server just like podman-remote would do?
Wouldn't this be more the task of something like auditd, fapolicyd or whatever tools exists for such a use case.
I cannot imagine that podman is special here, how would other applications deal with that?

Can you link to the bugzilla or were this was requested? 


---
Lastly this currently is super easy to circumvent. At the moment I could use `podman unshare` and then execute the commands inside this namespace they are not checked. Now if you say but they can block them from using podman unshare, you could do the same thing with nsenter and just set the correct env vars manually. This really does not feel secure.

**Comment by rhatdan** [comment]
The goal is to prevent accidental doing something you are not supposed to do, not dealing with malicious users.

Bottom line we have customers that were dealing with users using docker auth plugins, and wanted to block certain activity by users in their homedir.  They want similar functionality for rootless podman, and we offered this feature to them in order to meet their needs.  They agreed to the feature.  We can decide to document it in the future, but for now, I would just allow the script to print to stdout.

**Comment by Luap99** [comment]
> not dealing with malicious users.

Ok, but in this case why does the check need to happen in podman?
Couldn't you just do the same in a shell script wrapper then:

```
#!/bin/sh

# check "$@" here or execute whatever you like
exec /usr/bin/podman "$@"
```

Put this in `/usr/local/bin/podman` and make sure $PATH list this dir before `/usr/bin` and it should work.

**Comment by giuseppe** [comment]
I don't see any sane way to do it.  The user must block "podman unshare" with the same mechanism

/hold cancel

**Comment by giuseppe** [comment]
on a regular system there are so many ways to circumvent it that it doesn't even make sense to think about them.  What `podman unshare` does can be achieved with few lines of shell scripting; but this feature is what the customer needs

**Comment by Luap99** [comment]
> on a regular system there are so many ways to circumvent it that it doesn't even make sense to think about them. What `podman unshare` does can be achieved with few lines of shell scripting; but this feature is what the customer needs

yes that is what I don't really understand the purpose, to me this looks like a security feature but it really is not. If we aim for best effort allow them to limit non malicious users that is fine but it should be clear to anyone who is using this it is not a security feature. It is neither clear from the commit message nor from the code and there is no documentation so how should somebody know? To me the functionality and limitations should be documented. Just because this is a special request doesn't mean others shouldn't be allowed to use it IMO, podman is open source after all.

To me this looks like something that can easily be done with a shell script wrapper as mentioned above, you can certainly place this to `/usr/bin/podman` and move the actual podman binary somewhere else if this is required.

Either way if you really want this be merged you can easily outvote me. It is not like I have the right to block PRs when the majority of maintainers are happy with it.
The one things that really bothers me that it will be executed for `pkg/bindings` users? Could we check that `argv[0]` == `podman` or `podman-remote` to not do in such case? If the check is only for non malicious users anyway this shouldn't be a problem.

File: test/system/950-auth-scripts.bats
Comment (Luap99) [inline test/system/950-auth-scripts.bats:12]: If the customer wants to check podman-remote the test should be done for it as well.

File: pkg/rootless/rootless_linux.c
Comment (flouthoc) [inline pkg/rootless/rootless_linux.c:250]: nit / question: Should users be warned when if they unintentionally forgot to set their scripts as executable otherwise script silently skips them but its fine if this is expected behavior ?

**Review by flouthoc — APPROVED** [review]
LGTM just one question above.

File: pkg/rootless/rootless_linux.c
Comment (giuseppe) [reply pkg/rootless/rootless_linux.c:250]: no, the goal is to add readme files as part of that directory, so it is done intentionally

Comment (flouthoc) [reply pkg/rootless/rootless_linux.c:250]: SGTM

**Pull request merged**
